File: src/Model/ArrayType.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;

namespace AutoRest.Java.Model
{
    public class ArrayType : IType
    {
        public static readonly ArrayType ByteArray = new ArrayType(PrimitiveType.Byte, (string defaultValueExpression) => $"\"{defaultValueExpression}\".getBytes()");

        private ArrayType(IType elementType, Func<string,string> defaultValueExpressionConverter)
        {
            ElementType = elementType;
            DefaultValueExpressionConverter = defaultValueExpressionConverter;
        }

        public IType ElementType { get; }

        private Func<string, string> DefaultValueExpressionConverter { get; }

        public override string ToString()
        {
            return $"{ElementType}[]";
        }
        
        public IType AsNullable()
        {
            return this;
        }

        public bool Contains(IType type)
        {
            return this == type || ElementType.Contains(type);
        }

        public void AddImportsTo(ISet<string> imports, bool includeImplementationImports)
        {
            ElementType.AddImportsTo(imports, includeImplementationImports);
        }

        public string DefaultValueExpression(string sourceExpression)
        {
            string result = sourceExpression;
            if (result != null && DefaultValueExpressionConverter != null)
            {
                result = DefaultValueExpressionConverter(sourceExpression);
            }
            return result;
        }
    }
}

File: javagen/src/main/java/com/azure/autorest/model/clientmodel/ArrayType.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.autorest.model.clientmodel;

import java.util.Set;
import java.util.function.Function;


/**
 * The details of an array type that is used by a client.
 */
public class ArrayType implements IType {
    public static final ArrayType ByteArray = new ArrayType(PrimitiveType.Byte,
        defaultValueExpression -> defaultValueExpression == null
            ? "new byte[0]" // TODO (alzimmer): Should this be new byte[0] or null?
            : String.format("\"%1$s\".getBytes()", defaultValueExpression));

    private final String toStringValue;
    private final IType elementType;
    private final Function<String, String> defaultValueExpressionConverter;

    private ArrayType(IType elementType, Function<String, String> defaultValueExpressionConverter) {
        this.toStringValue = elementType + "[]";
        this.elementType = elementType;
        this.defaultValueExpressionConverter = defaultValueExpressionConverter;
    }

    public final IType getElementType() {
        return elementType;
    }

    @Override
    public String toString() {
        return toStringValue;
    }

    public final IType asNullable() {
        return this;
    }

    public final boolean contains(IType type) {
        return this == type || getElementType().contains(type);
    }

    public final void addImportsTo(Set<String> imports, boolean includeImplementationImports) {
        getElementType().addImportsTo(imports, includeImplementationImports);
    }

    public final String defaultValueExpression(String sourceExpression) {
        return defaultValueExpressionConverter.apply(sourceExpression);
    }

    @Override
    public String defaultValueExpression() {
        return defaultValueExpression(null);
    }

    public final IType getClientType() {
        // The only supported array type is byte[]
        return this;
    }

    public final String convertToClientType(String expression) {
        // The only supported array type is byte[]
        return expression;
    }

    public final String convertFromClientType(String expression) {
        // The only supported array type is byte[]
        return expression;
    }

    public String validate(String expression) {
        return null;
    }

    @Override
    public String jsonDeserializationMethod() {
        return "getBinary()";
    }

    @Override
    public String jsonFieldSerializationMethod(String jsonWriterName, String fieldName, String valueGetter) {
        return String.format("%s.writeBinaryField(\"%s\", %s)", jsonWriterName, fieldName, valueGetter);
    }

    @Override
    public String jsonValueSerializationMethod(String jsonWriterName, String valueGetter) {
        return String.format("%s.writeBinary(%s)", jsonWriterName, valueGetter);
    }

    @Override
    public String xmlAttributeSerializationMethod() {
        return "writeBinaryAttribute";
    }

    @Override
    public String xmlElementSerializationMethod() {
        return "writeBinaryElement";
    }
}
